controllers/helmchart_controller.go

```diff
@@ -481,32 +490,88 @@ func (r *HelmChartReconciler) buildFromHelmRepository(ctx context.Context, obj *
 			// Requeue as content of secret might change
 			return sreconcile.ResultEmpty, e
 		}
-	}
 
-	// Initialize the chart repository
-	chartRepo, err := repository.NewChartRepository(repo.Spec.URL, r.Storage.LocalPath(*repo.GetArtifact()), r.Getters, tlsConfig, clientOpts,
-		repository.WithMemoryCache(r.Storage.LocalPath(*repo.GetArtifact()), r.Cache, r.TTL, func(event string) {
-			r.IncCacheEvents(event, obj.Name, obj.Namespace)
-		}))
-	if err != nil {
-		// Any error requires a change in generation,
-		// which we should be informed about by the watcher
-		switch err.(type) {
-		case *url.Error:
-			e := &serror.Stalling{
-				Err:    fmt.Errorf("invalid Helm repository URL: %w", err),
-				Reason: sourcev1.URLInvalidReason,
+		// Build registryClient options from secret
+		logOpt, err := loginOptionFromSecret(*secret)
+		if err != nil {
+			e := &serror.Event{
+				Err:    fmt.Errorf("failed to configure Helm client with secret data: %w", err),
+				Reason: sourcev1.AuthenticationFailedReason,
 			}
 			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			// Requeue as content of secret might change
 			return sreconcile.ResultEmpty, e
-		default:
-			e := &serror.Stalling{
-				Err:    fmt.Errorf("failed to construct Helm client: %w", err),
-				Reason: meta.FailedReason,
+		}
+
+		logOpts = append([]registry.LoginOption{}, logOpt)
+	}
+
+	// Initialize the chart repository
+	var chartRepo chart.Remote
+	if repo.Spec.Type == sourcev1.HelmRepositoryTypeOCI {
+		if !registry.IsOCI(repo.Spec.URL) {
+			err := fmt.Errorf("invalid OCI registry URL: %s", repo.Spec.URL)
+			return chartRepoErrorReturn(err, obj)
+		}
+		// Tell the chart repository to use the OCI client with the configured getter
+		clientOpts = append(clientOpts, helmgetter.WithRegistryClient(r.RegistryClient))
+		ociChartRepo, err := repository.NewOCIChartRepository(repo.Spec.URL, repository.WithOCIGetter(r.Getters), repository.WithOCIGetterOptions(clientOpts), repository.WithOCIRegistryClient(r.RegistryClient))
+		if err != nil {
+			return chartRepoErrorReturn(err, obj)
+		}
+		chartRepo = ociChartRepo
+
+		// If login options are configured, use them to login to the registry
+		// The OCIGetter will later retrieve the stored credentials to pull the chart
+		if logOpts != nil {
+			// create a temporary file to store the credentials
+			// this is needed because otherwise the credentials are stored in ~/.docker/config.json.
+			// TODO@souleb: remove this once the registry move to Oras v2
+			// or rework to enable reusing credentials to avoid the unneccessary handshake operations
```

Comment:
The current credential store being used seems to be `.config/helm/registry/config.json` instead of `~/.docker/config.json`.
Even with the current code, that target store is still being used:

![image](https://user-images.githubusercontent.com/5452977/167883684-257dbfc9-2e58-4efa-94b2-db2513c7a49d.png)

Reply:
@stefanprodan  That's a very good point. Had a chat with @souleb on this - the E2E tests uses only public repositories, meaning that the authentication does not take place, and therefore the cred store is not created/amended.